### PR TITLE
Merge Cafebokning into master

### PR DESCRIPTION
**Pull request by davidwessman**
Preparing to launch Cafebokning

### CODE REVIEW

**Comment by Zofz** [comment]
Closes #74

File: app/assets/stylesheets/cafe_works.css.scss [hunk 1]
```diff
@@ -0,0 +1,21 @@
+@import "colors";
+
+.cafe-date {
+  background: $box-color;
+  border: 2px;
+  border-color: $border-color;
+  border-radius: 10px !important;
```
Comment (Zofz) [inline app/assets/stylesheets/cafe_works.css.scss]: Will have to be changed when the CSS is reworked.

File: app/controllers/contacts_controller.rb [hunk 2]
```diff
@@ -23,18 +23,19 @@ def new
   # GET /councils/1/edit
   def edit
   end
-  def mail    
-    if(params[:name]) && (params[:email]) && (params[:msg])
+
+  def mail
```
Comment (Zofz) [inline app/controllers/contacts_controller.rb]: Will rework in #87.

**Comment by Zofz** [comment]
@henrikssn 
When I try to pull down master to be up to date - it deletes all my cafe_work-files.
I guess this has something to do with the master having the ```revert```commit.
How do you solve that?

Other topic:
Can we delete this branch? https://github.com/fsek/web/tree/revert-73-revert-32-cafebokning/spec

**Comment by henrikssn** [comment]
What are you trying to do, `git merge master`?

**Comment by Zofz** [comment]
```git pull origin master```when on ```cafebokning```locally.
It just wiped my code away ;)

**Comment by henrikssn** [comment]
Hmm, that should work (and it works for me locally). Do you get any merge conflicts?

**Comment by Zofz** [comment]
Not any merge conflicts, it just deleted the code I was working on.
Do you still keep the ```controllers/admin/cafe_work``` for example?

**Comment by henrikssn** [comment]
I start to really like what I see here. Great job! :+1:

**Comment by davidwessman** [comment]
@henrikssn Thank you - Nice to hear 😄 

I will explain it, this can be used for the CarRent-module, the Event-signup module among others, making the model a lot thinner - I think 😄 

There are currently a lot of branches with the test. I have redone some work in multiple branches (all because I start to poke around to much when doing changes), hopefully we can get it together in a neat way 😄

**Comment by davidwessman** [comment]
@henrikssn @jforberg 
Any ideas why the:
```'/home/travis/build/fsek/web/vendor/bundle/ruby/2.1.0/cache/bundler/git/datetimepicker-rails-dbe0581d6e31538fee92ab883d261b9cacbf6108'```is failing?

**Comment by jforberg** [comment]
@davidwessman Link?

**Comment by davidwessman** [comment]
@jforberg It was a faulty gem, I managed to find it.

Now it is just the FactoryGirl.lint which I have solved locally.

File: app/models/cafe_work.rb [hunk 3]
```diff
@@ -0,0 +1,248 @@
+# encoding: UTF-8
+class CafeWork < ActiveRecord::Base
+  # Associations
+  belongs_to :profile
+  has_and_belongs_to_many :councils
```
Comment (davidwessman) [inline app/models/cafe_work.rb]: #76

**Comment by davidwessman** [comment]
Ready to merge if tests passes.
@fsek/admins

File: app/models/council.rb [hunk 4]
```diff
@@ -1,18 +1,29 @@
 # encoding: UTF-8
 class Council < ActiveRecord::Base
+  # Associations
   has_one :page, dependent: :destroy
   has_many :posts
   has_many :profiles, through: :posts
-  validates :title,:url,:presence => true
+  has_and_belongs_to_many :cafe_works
```
Comment (davidwessman) [inline app/models/council.rb]: #76

**Comment by henrikssn** [comment]
LGTM :+1:

**Comment by davidwessman** [comment]
You can look things through, we can merge later today if you have time.

File: app/models/cafe_work.rb [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+  # user is present.
+  # /d.wessman
+  def load(user)
+    self.attributes = worker.load_profile(user)
+  end
+
+  # Shows different status texts depending on the user.
+  # /d.wessman
+  def status_text(user)
+    case status_view(user)
```
Comment (henrikssn) [inline app/models/cafe_work.rb]: You need to use literals here instead of integers. It is impossible to understand what status 0/1/2/3 means otherwise.

Comment (davidwessman) [inline app/models/cafe_work.rb]: Will draft something up tomorrow 😄

Comment (davidwessman) [inline app/models/cafe_work.rb]: How should I do then? Should I return a symbol or something then? 
Like 
- :sign_up
- :edit
- :authorize
- :assigned
? 😄 @henrikssn

Comment (henrikssn) [inline app/models/cafe_work.rb]: Yes, that would be a much better solution :+1:

**Comment by davidwessman** [comment]
@jforberg - I tried to move from HABTM to a join model.
When I create ```cafe_work_council```the table in the database will be called ```cafe_work_councils```.
The HABTM-join-table for the same relation would be called ```cafe_works_councils``` and have ``ìd: false```.

How should we fix this on the production?

**Comment by jforberg** [comment]
@davidwessman There are two ways.

1. Rename the table.
2. Tell the model to use a different table with ``self.table_name 'some_table'

I'd go with (2) but it's important to be consistent later.

As for the id you need to do it in three steps:

1. ALTER TABLE to add the ID column which has to be an ordinary column for now.
2. Iterate through the existing rows and add ids. This has to be done in raw SQL because, as stated, HABTM gives us no way to differentiate between rows on the framework level.
3. ALTER TABLE to make the ID column the primary key.
4. Optional remove duplicates. This can be done in SQL or the framework

**But the id part can wait**, the important thing is to move away from HABTM in the framework. Rails **can** actually deal with tables that don't have ID columns. It's just nicer in the long run.

**Comment by davidwessman** [comment]
@fsek/admins I ran into some issues with setting up the Cancancan.
I am unable to even render the Councils controller when having the ```load_permissions_and_authorize_resource```.
Do I need to do the seed action first?

**Comment by henrikssn** [comment]
You need to have permissions to see it. Try this in you console
`p = Permission.create!(subject_class: :all, action: :manage)`
`post = Post.create!(title: 'Admin', description: 'admin')`
`post.permissions << p`
`User.find_by_id(your_user_id).profile.posts << post`

File: app/controllers/posts_controller.rb [hunk 6]
```diff
@@ -74,20 +69,18 @@ def create
     end
   end
 
-  # PATCH/PUT /news/1
-  # PATCH/PUT /news/1.json
-  def update    
+  def update
```
Comment (davidwessman) [inline app/controllers/posts_controller.rb]: Needs complete rework.

File: app/controllers/posts_controller.rb [hunk 7]
```diff
@@ -2,15 +2,15 @@
...
-  before_action :set_post, only: [:show, :edit, :update, :destroy, :remove_profile,:add_profile_username, :display]
 
   def remove_profile
     @profile = Profile.find_by_id(params[:profile_id])
-    @post.profiles.delete(@profile)    
+    @post.profiles.delete(@profile)
     respond_to do |format|
-      format.html { redirect_to council_posts_path(@council), notice: @profile.name.to_s + ' har inte längre posten ' + @post.title.to_s + '.'}
-    end 
+      format.html { redirect_to council_posts_path(@council), notice: @profile.name.to_s + ' har inte längre posten ' + @post.title.to_s + '.' }
```
Comment (davidwessman) [inline app/controllers/posts_controller.rb]: Needs complete rework.

File: app/controllers/posts_controller.rb [hunk 8]
```diff
@@ -18,53 +18,48 @@ def add_profile_username
     end
     if @profile == nil
       respond_to do |format|
-        format.html { redirect_to council_posts_path(@council), flash: {alert: 'Hittade ingen användare med det användarnamnet.'}}
+        format.html { redirect_to council_posts_path(@council), flash: {alert: 'Hittade ingen användare med det användarnamnet.'} }
```
Comment (davidwessman) [inline app/controllers/posts_controller.rb]: Needs complete rework.

**Comment by davidwessman** [comment]
@fsek/admins 
Ready to merge this with master (or stage)
I implemented lots of Cancancan code into different connected controllers, so please have a look at that as well.

**Comment by davidwessman** [comment]
Will close:
- #137 
- #134 
- #133 (was also fixed in #135, but changed here)
- #121 (should be increased - and scoped for only ```development```)

**Comment by davidwessman** [comment]
Merged #87 into this and closed it.
Closes #72

**Comment by davidwessman** [comment]
Fixes #138 with a50c281

**Comment by henrikssn** [comment]
I added some comments to a909eb7, please fix those before you merge. In general, take a look at `spec/controllers/election/candidates_controller_spec.rb` for a "how-to" on writing controller tests with cancan

**Comment by henrikssn** [comment]
I added a short descriptions for controller tests at https://github.com/fsek/web/wiki/How-to-write-a-controller-test

**Comment by davidwessman** [comment]
@henrikssn Nice wiki! Now it makes more sense 😄 
I fixed all you comments (I think).
Feeling ready to merge.

**Comment by davidwessman** [comment]
@fsek/admins Please have a look so that we can merge tonight 😄

**Comment by davidwessman** [comment]
Fixed #132

**Comment by henrikssn** [comment]
Great work! Fix the last comments and then you have LGTM from me :)

File: spec/controllers/admin/cafe_works_controller_spec.rb [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+  let(:not_owner) { create(:user) }
+  let(:cwork_profile) { create(:cafe_work, :w_profile, profile: user.profile) }
+  let(:cwork_access) { create(:cafe_work, :access) }
+  let(:cwork) { create(:cafe_work) }
+
+  # Hack to become admin
+  # Should be changed when Cancancan is implemented
+  # /d.wessman
+  before(:each) do
+    sign_in user
```
Comment (henrikssn) [inline spec/controllers/admin/cafe_works_controller_spec.rb]: Try to replace this with `allow(controller).to receive(:currrent_user).and_return(user)`

**Comment by davidwessman** [comment]
@henrikssn The Posts-controller needs a total rework, I won't do anything about it now.
The css !important thing will have to wait until the CSS is reworked, otherwise it won't show.
I cannot make the ones with assignment branch too big much smaller.

**Comment by SirDirac** [comment]
This PR was deployed to Production. Reference: <a href='http://samson.fsektionen.se/projects/web/deploys/71' target='_blank'>fe22738</a>

**Comment by davidwessman** [comment]
@henrikssn Something did not work -.-
Why is the server not logging properly?

**Comment by henrikssn** [comment]
Logs should be found at `/home/deploy/production/current/log/production.log`

Try deploying at dev server.

**Comment by davidwessman** [comment]
Aha.
Do we still hard reset dev?

**Pull request merged**